Qty export multiple files

falken10vdl opened the pull request:
This is a PR following conversation https://community.osarch.org/discussion/2864/csv-export-from-multiple-models-future-patch#latest
It allows to export data from multiple files
The relevant documentation is also added.
Thanks!

Ghesselink committed:
parsed mvd description in ifcopenshell.file

Ghesselink committed:
handle emptry description field

Ghesselink committed:
abort string when None, more explicit error handling

Ghesselink committed:
Allow for dots in mvd value

falken10vdl committed:
Export multiple ifc

Moult commented:
Wow this is a big PR. A huge thanks! I haven't had a chance to look in detail, but it looks as though you've created a new way to add linked models. Why don't you just reuse the existing Links panel / feature? Or have I misunderstood?

falken10vdl commented:
Yes, @steverugi suggested to put maybe a tick to take care of the linked files. I just felt that maybe one wanted to see in the same place all the files that one wants to use for. Qty to spreadsheet export (and the ability to select maybe some but not all). I was also thinking that if you have different LoD you might want to add files which are not linked so a UIList open to whatever one wants. Both for linked and not linked files.
Thanks!

Moult commented:
I'd prefer to keep it part of the "Links" system, but perhaps allow for "dormant" (? is there a better word) links. Or maybe two types of links, geometric links and data links?

A link (either geometric or data) can then be selectively enabled or disabled for IfcCSV, but there is a master list of links. This way it can also be reused for things like IfcFM, IfcClash, search filters across models, drawing visibility, IfcTester, etc, with reusable UI.

falken10vdl commented:
Ok. I am not sure I am following you :). The "Add linked files" is not doing anything except copying the filepaths from the links section for convenience. The files in the UIList are then treated the same way. They are processed very much like single files one after another.

maxfb87 committed:
Purge unecessary code

falken10vdl commented:
After reading a little bit more maybe I understand better. You want to have the links section maybe adding enable / disable for csv export (or other actions as you say). An then keep the non liked files in the import/export panel adding a tick to take in/or not account linked files? I can update the PR like that if you guys think it is better. Let me know. Thanks!

Ghesselink committed:
mvd info to module, add testing, setattr & getattr

Moult committed:
Fix #6501. Fix #6104. Fix #6189. You can now change wall offset convention after drawing walls.

People kept on getting confused with the align tool thinking it changed
the baseline to reference line offset whereas it actually just aligned
the object bodies. This tool now does exactly that.

Also start refactoring the "DumbWall" classes into the tools so we can
test them properly.

Andrej730 committed:
Use isodate to parse iso durations #2783

Related #6585

Co-Authored-By: sboddy <12862430+sboddy@users.noreply.github.com>

Andrej730 committed:
typing

Andrej730 committed:
black .

Andrej730 committed:
bonsai Makefile - specify shell explicitly

To avoid Windows trying to use cmd

Andrej730 committed:
Bump IOS build in Bonsai

New build resolves #6529 #6576

maxfb87 committed:
Add calculating functions for earthworks

aothms committed:
Debug build_win.yml

Andrej730 committed:
bim.assign_document - add to documents UI

Andrej730 committed:
ifc5d - use filename for imported schedule name

So it will be in sync with how export works.

Removed non_sor_fields is unrelated - forgot to remove before.

Andrej730 committed:
ifc5d - strip " Cost" suffix when importing csv

" Cost" is a suffix is added on export, so it makes to get rid of it on import.

Andrej730 committed:
Link.name - fix description

After f8828d3 we store linked filepath in .ifc project, so we don't support path relative to .blend anymore.

Moult committed:
See #2783. Duration parsing no longer uses isodate to remain faithful to formatting.

steverugi commented:
@falken10vdl 
do you think it's possible to have a mockup of what the final panel is going to look like? a snapshot would do I guess
Sorry I lost track of the changes we want to implement
thanks

Andrej730 committed:
tool.Ifc - add some doc-strings

Moult committed:
See #2783. Don't show 0 durations for clarity.

Andrej730 committed:
Fix issue with custom diagram scale (1d5891f)

It was referring to non-existent prop "custom_diagram_scale"

Andrej730 committed:
readable_ifc_duration - add doc-string and note

Andrej730 committed:
black .

falken10vdl commented:
@steverugi Currently the PR has the panel as given in the documentation (see attached pdf)
[Spreadsheet Export - Bonsai 0.8.2 documentation.pdf](https://github.com/user-attachments/files/19835102/Spreadsheet.Export.-.Bonsai.0.8.2.documentation.pdf)

So the linked files part is as follows:
<img width="518" alt="image" src="https://github.com/user-attachments/assets/41b5d1dc-5feb-4f30-bf79-de1493fd2fc6" />
I simply copy the pathfiles from one panel to the other for convenience. Nothing else.
Anyhow if that is creating confusion we can adapt it as you think is easiest to understand.
Please note that my mother tongue is Spanish so maybe my words are not the best to be clearly understood :)

steverugi commented:
Hola @falken10vdl :)

OK so in practice the already linked models are brought in the panel where we can select files to be used in the new Spreadsheet I/O, allowing the input from multiple models instead of one as it is now. 
I don't personally see any confusion there.
It's perfect, looking forward to using it.

Thanks a lot for your help

Ghesselink committed:
set set/get_description, add view_definitions & tests

falken10vdl commented:
Correcto :) Yes, it is just for convenience. Nothing is touched from the logic/handling of linked models. It is just that in spreadsheet I/O one can select whatever files one wants with the file explorer to use for exporting. One could select really files not related to the current ifc file and I was just thinking that since we already have the paths for the linked models, with the push of a button we can populate the list.
Thanks!

Andrej730 committed:
typing

Andrej730 committed:
Add target_view prop to camera

This property was already set in 50b41f5 but Blender property was missing. Though nowit's not very useful as it just writes value to pset.

Andrej730 committed:
Fix error linking ifc files with relative filepaths going upwards

Mentioned in https://github.com/IfcOpenShell/IfcOpenShell/pull/6540#issuecomment-2820938852

Andrej730 committed:
typing

Andrej730 committed:
bim.override_mesh_separate - use original Blender enum and description

Andrej730 committed:
bim.override_mesh_separate - invoke menu if called without argument

Mimicking default mesh.separate behaviour.

@Gorgious56 I don't know if you've seen this before, but you might be interested - I've found a way to figure if operator wasn't called without any arguments.

Andrej730 committed:
bim.override_mesh_separate - use operator_enum, hide Blender Separate

as it's now integrated in IFC separate and using Blender Separate for ifc elements is not safe

Andrej730 committed:
bim.override_origin_set - use operator_enum for menu

Andrej730 committed:
bim.override_mesh_separate - integrate mesh.separate

mesh.separate will be used for all other selected non-ifc objects.

Included some code to support multple objects but there are still a few issue with it, I'll look into soon.

Andrej730 committed:
bim.override_mesh_separate - fix bug separating elements (7bc4b51)

`obj` variable was used but never passed to `separate_element`

```
Error: Python: Traceback (most recent call last):
  File "\bonsai\bim\ifc.py", line 458, in execute_ifc_operator
    result = getattr(operator, "_execute")(context)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "\bonsai\bim\module\geometry\operator.py", line 105, in _execute
    self.separate_element(element)
  File "\bonsai\bim\module\geometry\operator.py", line 175, in separate_element
    new_objs = [obj]
                ^^^
NameError: name 'obj' is not defined
```

Andrej730 committed:
bim.override_mesh_separate - fix bug with skip_save

Previously you would use bim.override_mesh_separate from UI with any of the available `type` options, then running operator separately would leading to running this operator from F3 to always choose this options (possibly leading to errors in objects mode)

Andrej730 committed:
bim.override_mesh_separate - handle possible mode mismatch more gracefully

Other options LOOSE_PARTS and MATERIAL do support object mode.

Andrej730 committed:
bim.expand_target_view - remove trailing space from description

Andrej730 committed:
Fix invalid cprops after camera data is recreated

Mentioned in  #5676, discoverd after 476604202a.

`bim.update_representation` removes previous `camera.data` and adds a new one making cprops invalid possibly leading to either invalid prop values or crashes (https://projects.blender.org/blender/blender/issues/137889).

Andrej730 committed:
typing

Andrej730 committed:
Fix resaving camera to IFC on every drawing activation (32899eb)

`camera_props.update_representation` basically always returned True leading to `bim.update_representation` call on camera on every drawing activation. It was always True because `create_camera` wasn't setting `representation` and some other related camera props.

Changes:
- create_camera now setting representation
- ortho_scale, rather_x and raster_y now explicitly is tied to width, height and drawing scale props, not on depsgraph update. Depsgraph update wasn't allowing to set these props in time for create_camera to record representation correctly
- depsgraph update now only in charge of ortho_scale (just to be sure user won't touch it by hand) and scene.render resolution

Andrej730 committed:
create_camera - small optimization with early return

Andrej730 committed:
create_camera - small optimization with util.shape

Andrej730 committed:
IFC Mesh Separate - add to object mode menu too for discoverability + support multiple selected elements

Short demo - https://imgur.com/a/PIuJ9Rq

Andrej730 committed:
core tests - more readable errors

E.g. it shows now:
AttributeError: Interface 'bonsai.core.tool.Drawing' has no attribute 'weird_method'.

Instead of:
Prophecy <class 'abc.Drawing'> has no attribute weird_method

Andrej730 committed:
typing

maxfb87 committed:
Fix bug in reload link ifc where in Windows gives an error if relative path is used

Andrej730 committed:
core tests - very readable error for unpredicted return values

Previously:
E       TypeError: Object of type Prophecy is not JSON serializable

Now:
E           TypeError: Failed to serialize call: '{'name': 'duplicate_ifc_objects', 'args': ([<Prophecy for 'bonsai.core.tool.Ifc'>],), 'kwargs': {}}'.
E           Possibly due to unpredicted return value for some call.
E           See the list of the recent calls without return values:
E           - {'name': 'get_object', 'args': ('annotation',), 'kwargs': {}}
E           - {'name': 'run', 'args': ('group.assign_group',), 'kwargs': {'group': 'new_group', 'products': ['new_drawing']}}
E           - {'name': 'run', 'args': ('group.edit_group',), 'kwargs': {'group': 'new_group', 'attributes': {'Name': 'unique_name', 'ObjectType': 'DRAWING'}}}
E           - {'name': 'run', 'args': ('group.unassign_group',), 'kwargs': {'group': 'group', 'products': ['new_drawing']}}

Andrej730 committed:
Fix core test for drawing duplication (0b1c049)

Andrej730 committed:
core.duplicate_drawing - remove debug print statements and unused code

Andrej730 committed:
IFC Mesh Separate - reload representation just once if multiple items got separated

Andrej730 committed:
IFC Mesh Separate - fix bug when separating an item wouldn't change separated item

Mentioned by Ryan in #6285

Andrej730 committed:
entity_instance.__getattr__ - document some attributes conversion to Python

Andrej730 committed:
IFC Mesh Separate - fix bug when new separated mesh wasn't relinked to a new ifc item

Andrej730 committed:
bim.import_representation_items - add description

Andrej730 committed:
IFC Mesh Separate - fix not updated items decorator after some item was separated

Noticed fixing bug from #6285

Andrej730 committed:
IFC Mesh Separate - fix bug in 52af302bef separating multiple elements

Forgot to clear selection after processing each object, it could have led to incorrect results since mesh.separate is applied to the selection later.

Andrej730 committed:
Revert "Bump IOS build in Bonsai" #6589

This reverts commit 50e7728f95e8fa36193303c51c3d58725be33fa9.

c4rlosdias committed:
clear list of entities failed when execute Tester

theoryshaw committed:
Fix #6601 - If you duplicate something it makes the duplicate the active object.
also bulk deselect all objects_to_remove to speed things up a little.

theoryshaw committed:
small tweak to previous commit

theoryshaw committed:
whoops

Andrej730 committed:
IFC Join - add poll check similar to object.join

Andrej730 committed:
typing

Andrej730 committed:
IFC Join - add description

Andrej730 committed:
IFC Join - prevent possible ghost ifc objects after join

Skip ifc objects of different type as they will be ignored by object.join either way and removing them produced ghost ifc objects.

Andrej730 committed:
IFC Join - prevent joining rep items into non items

Because it was removing representation item object leaving Bonsai in confusing state, which may lead to errors later

Andrej730 committed:
IFC Join - fix error joining representation items (6ce4b5d)

It was trying to reaccess already removed ifc_definition_id leading to an error. Now we update id with a new item.

Andrej730 committed:
geometry.name_item_object

Andrej730 committed:
reuse ifc.link for meshes

Andrej730 committed:
edit_meshlike_item - fix missing item object name update

Andrej730 committed:
IFC Join - support joining non-ifc Blender objects into representation items

Short demo - https://imgur.com/a/o9MJiIr

brunopostle committed:
Linux escape double-quotes in filenames #6630

Special chars ('#{)$ were already escaped, but " wasn't

aothms committed:
Mark settings as internal #6623

aothms committed:
Don't set xstep.cascade.unit as it appears to result in scaling during export #6623

Andrej730 committed:
get_uri, resolve_uri - refactor #6605

1) Use pathlib for consistency
2) Ensure consistent output - always use posix format and resolve paths (but keep the symlinks)

Andrej730 committed:
Fix issues printing drawing with projects linked by relative path

The fix in f9bf889 wasn't related to this problem and was targetting another bug.

Andrej730 committed:
Fix missed cases of loading unloaded linked files in 9316e0b (#6403)

Andrej730 committed:
typing

Andrej730 committed:
Revert f9bf889 #6605

As this fix is no longer needed and covered by more general resolve_uri refactor

Andrej730 committed:
import_ifc - skip setup_arrays for annotations #6636

Andrej730 committed:
Setup array constraints for annotations that use paramatric arrays

Previously you was able to add parametric array to annotations but constraint would be missing on file reload.

Andrej730 committed:
bim.add_array - guard from unexpected element types

So we can gradually expand it with the things we do know to support.

Andrej730 committed:
mesh.add_grid - display props in Blender project units

Example - https://i.imgur.com/koLiH7U.png

Andrej730 committed:
Use actual text position instead of local pos rendering drawings text decorations

E.g. it's important if text is part of parametric array and it's local position may be constant, but it's constrainet by parent's position.

aothms committed:
One more self-intersection check in cgal kernel #6589

aothms committed:
todo notes #6589

aothms committed:
partial_success_is_success #6589

Andrej730 committed:
Operator to run git diff on current ifc file

Short demo - https://imgur.com/a/TvnztNy

Andrej730 committed:
create_axis_curve - fix typo

Andrej730 committed:
ci-black-formatting - check syntax errors explicitly

FabioPiccinini committed:
Use tool.Blender.get_view3d_area() to allow the use of separate windows (#6617)

aothms committed:
partial_success_is_success #6589

Andrej730 committed:
typing

Andrej730 committed:
IfcGrid - add placement by default to avoid invalid IFC

Andrej730 committed:
Fix invalid IfcGrid cartesian points (774a770)

They were saved as 4d coordinates, not 2d.

Andrej730 committed:
black .

aothms committed:
Fix faceset helper on advanced brep oriented edges #6604

aothms committed:
Defensiveness against empty faces

aothms committed:
Bounds checking in variant array

Andrej730 committed:
Fix typo

Andrej730 committed:
More consistent Windows Bonsai .ifc file association

Rely on `winreg` instead of running powershell scripts with cmd commands + .reg file + weird workflow with the need to retype cmd command. Also use HKCU instead of HKCR because it doesn't require admin rights.

If someone hearing about this feature for the first time, here's a short demo - https://imgur.com/a/dfG44pY (icon appears too but after explorer restart).

Andrej730 committed:
Windows - fix opening paths with special characters #6630

Andrej730 committed:
Fix bug creating a drawing if linked model schema doesn't match the main model schema

Co-Authored-By: Thomas Krijnen <1096535+aothms@users.noreply.github.com>

Andrej730 committed:
Reenable black formatting workflow

Since 3b1e893 black formatter workflow basically had no use since it wasn't reporting any issues (no one checks workflow logs unless workflow failed) and wasn't reliable for catching syntax errors either (black formatter catches syntax errors only when they get in the way of it's formatting the code, but now we delegate to python itself to catch syntax errors, see 406962287c).

And we also have some docs on how to apply black formatter - https://docs.bonsaibim.org/guides/development/code_style.html

Andrej730 committed:
new operator - bim.select_entity_by_guid

I find myself often having element guid copied (either from other viewer or from text editor or from elsewhere) and to find it in scene I needed to create a new search group, add a filter, paste guid and then search for it. So decided to add a shortcut for this that just takes guid from the clipboard and runs the search for it.

Location - https://i.imgur.com/3c5rm1N.png

Andrej730 committed:
Reuse get_last_commit_hash

Noticed that hash wasn't updating from live git repo because of this.
Also update commit data only when running the package, not extension.

Andrej730 committed:
Display actual Bonsai version for live repo (reuse last commit date)

Andrej730 committed:
color by property - add operators descriptions

Andrej730 committed:
typing

Andrej730 committed:
Fix ifc2x3 errors editing IfcSurfaceStyleShading #6624

Andrej730 committed:
Fix dangerous bug with bim.override_outliner_delete

Essentialy outliner.delete should do the same as object.delete, but in our case the main implementation was in object.delete while outliner.delete was missing very important parts - e.g. handling arrays.

So if user could have remove part of an array without errors, leaving pset data invalid. See #6590, #6547.
Found this by trying to debug issues above.

Now outliner.delete will call object.delete explicitly, so the two stay connected.

I also had an idea that may be it somehow can produce #6632/#6619, but couldn't find a way to reproduce it. So it's still uncertain.

Andrej730 committed:
overridden delete operators - add descriptions

Ghesselink committed:
validate ifcopenshell.simple_spf.file

Ghesselink committed:
handle header entity string reversal

Andrej730 committed:
Bonsai docs - add page to list hotkeys

See https://docs-unstable.bonsaibim.org/reference/hotkeys.html

Andrej730 committed:
typing

Andrej730 committed:
Fix bim.override_move issue with aggregates #6648 (c9fc543)

falken10vdl committed:
Skip non MESH and CURVE object listing elements for coloring #6657

So non MESH or CURVE objects are not listed in the Colour by property.

Now all unrelated empty objects are excluded from the coloring section as they were just taking space, see - https://imgur.com/sPA8I5V

Andrej730 committed:
IFC delete - just reuse object.delete if no IFC file active

So it will be possible for this operator to completely override object.delete

Andrej730 committed:
Override Shift-X and Shift-Del with IFC Delete

To prevent users from creating weird state of IFC model by using non-ifc `object.delete` triggered by those hotkeys by default.

Ghesselink committed:
Valid template header default (#6610)

* Valid template header default

* consistent version in tempalte file

* schema-dependent mvd

aothms committed:
validate.py be more specific in distinguishing type decl and entity refs

aothms committed:
validate.py Add test files

Andrej730 committed:
Use mesh.separate for vanilla Blender objects #6658 (0b22333)

mesh.separate works completely different in EDIT mode, previously we only supported it from object mode

Andrej730 committed:
black .

Ghesselink committed:
use repr(), compatible f.header

Ghesselink committed:
Update src/ifcopenshell-python/ifcopenshell/validate.py

Co-authored-by: Thomas Krijnen <t.krijnen@gmail.com>

Ghesselink committed:
hasattr to isinstance()

Ghesselink committed:
header_entity type -> attr check

aothms committed:
NB: fix get_cartesiantransformationoperator3d()

Andrej730 committed:
black .

Andrej730 committed:
bim.override_move - rename for clairty and add description

Andrej730 committed:
Fix error duplicating just an IfcElementAssembly (d2b05ea)

Example error traceback:
```
Traceback (most recent call last):
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\bonsai\bim\module\geometry\operator.py", line 1044, in execute
    return OverrideDuplicateMove.execute_duplicate_operator(self, context, linked=False)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\bonsai\bim\module\geometry\operator.py", line 1055, in execute_duplicate_operator
    IfcStore.execute_ifc_operator(operator, context)
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\bonsai\bim\ifc.py", line 458, in execute_ifc_operator
    result = getattr(operator, "_execute")(context)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\bonsai\bim\module\geometry\operator.py", line 1047, in _execute
    return OverrideDuplicateMove.execute_ifc_duplicate_operator(self, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\bonsai\bim\module\geometry\operator.py", line 1085, in execute_ifc_duplicate_operator
    old_to_new, new_active_obj = tool.Geometry.duplicate_ifc_objects(
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\bonsai\tool\geometry.py", line 2131, in duplicate_ifc_objects
    tool.Root.recreate_aggregate(old_to_new)
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\bonsai\tool\root.py", line 397, in recreate_aggregate
    tool.Blender.select_and_activate_single_object(bpy.context, tool.Ifc.get_object(new_aggregate[0]))
                                                                                    ~~~~~~~~~~~~~^^^
TypeError: 'NoneType' object is not subscriptable
```

Andrej730 committed:
bim.add_aggregate - update description

Andrej730 committed:
bim.add_aggregate - remove unused 'obj' property (c51aa91)

Andrej730 committed:
bim.add_aggregate - use more common way to get all selected objects

Andrej730 committed:
aggregate decorator indicator - make it more intuitive

Example - https://imgur.com/a/RoFQKa1

Andrej730 committed:
typing

Andrej730 committed:
object.delete, outliner.delete - move to common method is_batch logic

Andrej730 committed:
outliner.delete - make errors from attempt to delete locked object less scary

Traceback errors give users an impression that some broke and maybe project now is in some invalid state and they should ctrl-z immediately, error reports are much more friendly.

Before - https://i.imgur.com/N4fUU1M.png
After - https://i.imgur.com/6On1Be0.png

Andrej730 committed:
Fix bim search tests after 0b22333

Andrej730 committed:
bonsai pytest.ini - specify plugins explicitly

To make it possible to run the tests with PYTEST_DISABLE_PLUGIN_AUTOLOAD set, which is very useful if you don't want to start Blender anytime you run any tests on your system.

Andrej730 committed:
add pytest-bdd to bonsai daily cli

Andrej730 committed:
Fix mistake in fd30a6b #6670

Andrej730 committed:
Fix constant camera representation update because of the float garbage values

Similar issue to 82f25f5

Basically matrix is a little bit different every time you activate drawing even if user made no changes to camera position. My guess is basically all float values stored by Blender somewhere deep as float32 but when you access matrix world, Blender converts them to Python floats (which are float64) and some garbage values introduced along the way creating this noise.

Andrej730 committed:
add_representation - more readable error on using panoramic camera

Andrej730 committed:
typing

Andrej730 committed:
Expose camera type to UI and automatically update IFC camera type on drawing activation

Noticed fixing #6670 that though example project in this issue had camera named "PERSPECTIVE" but it was saved to IFC as orthogonal and there is no straightforward way to change it, besides changing Blender camera props and updating representation.

Example - https://imgur.com/a/GQ9A3EK

It's kind of exposed to UI now, but there's still bunch of issues with it to address.

Andrej730 committed:
Document features available for perspective camera in it's description

See https://i.imgur.com/8ZPJ8ag.png

Ghesselink committed:
update latest simple_spf submodue

Andrej730 committed:
Fix missing IFC undo for all activate drawing operations

Very dangerous bug that could create a situation when .blend will go out of sync with IFC leading to unpredictable issues.
Example issue:
1. Create a drawing and activate it.
2. Create a grid.
3. Activate the drawing again.
4. Undoing just drawing activation here would work fine - annotations would be still floating in IFC without objects, but that would be fixed on the next drawing activation

To make it more dangerous undo grid creation. Then you'll get the traceback similar to the one below. The most dangerous part of it, that user won't see an error, it will just pop up in the system console silently.
```
Traceback (most recent call last):
  File "\bonsai\bim\handler.py", line 244, in undo_post
    IfcStore.undo(until_key=props.last_transaction)
  File "\bonsai\bim\ifc.py", line 533, in undo
    transaction["rollback"](transaction["data"])
  File "\bonsai\bim\ifc.py", line 449, in <lambda>
    operator, rollback=lambda d: tool.Ifc.get().undo(), commit=lambda d: tool.Ifc.get().redo()
                                 ^^^^^^^^^^^^^^^^^^^^^
  File "\ifcopenshell\file.py", line 345, in undo
    transaction.rollback()
  File "\ifcopenshell\file.py", line 161, in rollback
    element = self.file.by_id(operation["id"])
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "\ifcopenshell\file.py", line 498, in by_id
    return self[id]
           ~~~~^^^^
  File "\ifcopenshell\file.py", line 485, in __getitem__
    return entity_instance(self.wrapped_data.by_id(key), self)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "\ifcopenshell\ifcopenshell_wrapper.py", line 8884, in by_id
    return _ifcopenshell_wrapper.file_by_id(self, id)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Instance 1565 not found
```

Andrej730 committed:
ifcopenshell.file - save last failed transaction for introspection

Andrej730 committed:
Bonsai - make undo errors visible to user

In theory there are possible undo errors, when some step wasn't registered in undo system and undo then won't be able to undo change from IFC. The problem is, those errors could be very dangerous but they occur in background, undetected by users, they're only visible as console traceback.

After this commit we'll try to detect them and set `last_error`, which will show usual Bonsai error panel in Properties.

@theoryshaw just in case, as you will probably find this kind of error 😁

Example traceback user might see in system console (or in their error report from "Copy Error Message").
```
last_error: Undo transaction operations:
- {'action': 'create', 'value': {'id': 1560, 'type': 'IfcGrid', 'GlobalId': '05Y8D_Ja12HfiEbmc$D3WN', 'OwnerHistory': None, 'Name': None, 'Description': None, 'ObjectType': None, 'ObjectPlacement': None, 'Representation': None, 'UAxes': None, 'VAxes': None, 'WAxes': None, 'PredefinedType': None}}
- {'action': 'edit', 'id': 1560, 'index': 2, 'old': None, 'new': 'Grid'}
- {'action': 'edit', 'id': 1221, 'index': 4, 'old': ({'id': 1297}, {'id': 1247}, {'id': 1220}, {'id': 1272}), 'new': ({'id': 1272}, {'id': 1247}, {'id': 1220}, {'id': 1560}, {'id': 1297})}
- {'action': 'create', 'value': {'id': 1561, 'type': 'IfcCartesianPoint', 'Coordinates': (0.0, 0.0, 0.0)}}
- {'action': 'create', 'value': {'id': 1562, 'type': 'IfcDirection', 'DirectionRatios': (0.0, 0.0, 1.0)}}
- {'action': 'create', 'value': {'id': 1563, 'type': 'IfcDirection', 'DirectionRatios': (1.0, 0.0, 0.0)}}
- {'action': 'create', 'value': {'id': 1564, 'type': 'IfcAxis2Placement3D', 'Location': {'id': 1561}, 'Axis': {'id': 1562}, 'RefDirection': {'id': 1563}}}
- {'action': 'create', 'value': {'id': 1565, 'type': 'IfcLocalPlacement', 'PlacementRelTo': None, 'RelativePlacement': {'id': 1564}}}
- {'action': 'edit', 'id': 1565, 'index': 0, 'old': None, 'new': {'id': 66}}
- {'action': 'edit', 'id': 1560, 'index': 5, 'old': None, 'new': {'id': 1565}}
- {'action': 'create', 'value': {'id': 1566, 'type': 'IfcGridAxis', 'AxisTag': 'A', 'AxisCurve': None, 'SameSense': True}}
- {'action': 'edit', 'id': 1560, 'index': 7, 'old': None, 'new': ({'id': 1566},)}
- {'action': 'create', 'value': {'id': 1567, 'type': 'IfcCartesianPoint', 'Coordinates': (-2.0, 0.0)}}
- {'action': 'create', 'value': {'id': 1568, 'type': 'IfcCartesianPoint', 'Coordinates': (22.0, 0.0)}}
- {'action': 'create', 'value': {'id': 1569, 'type': 'IfcPolyline', 'Points': ({'id': 1567}, {'id': 1568})}}
- {'action': 'edit', 'id': 1566, 'index': 1, 'old': None, 'new': {'id': 1569}}
- {'action': 'create', 'value': {'id': 1570, 'type': 'IfcGridAxis', 'AxisTag': 'B', 'AxisCurve': None, 'SameSense': True}}
- {'action': 'edit', 'id': 1560, 'index': 7, 'old': ({'id': 1566},), 'new': ({'id': 1566}, {'id': 1570})}
- {'action': 'create', 'value': {'id': 1571, 'type': 'IfcCartesianPoint', 'Coordinates': (-2.0, 10.0)}}
- {'action': 'create', 'value': {'id': 1572, 'type': 'IfcCartesianPoint', 'Coordinates': (22.0, 10.0)}}
- {'action': 'create', 'value': {'id': 1573, 'type': 'IfcPolyline', 'Points': ({'id': 1571}, {'id': 1572})}}
- {'action': 'edit', 'id': 1570, 'index': 1, 'old': None, 'new': {'id': 1573}}
- {'action': 'create', 'value': {'id': 1574, 'type': 'IfcGridAxis', 'AxisTag': 'C', 'AxisCurve': None, 'SameSense': True}}
- {'action': 'edit', 'id': 1560, 'index': 7, 'old': ({'id': 1566}, {'id': 1570}), 'new': ({'id': 1566}, {'id': 1570}, {'id': 1574})}
- {'action': 'create', 'value': {'id': 1575, 'type': 'IfcCartesianPoint', 'Coordinates': (-2.0, 20.0)}}
- {'action': 'create', 'value': {'id': 1576, 'type': 'IfcCartesianPoint', 'Coordinates': (22.0, 20.0)}}
- {'action': 'create', 'value': {'id': 1577, 'type': 'IfcPolyline', 'Points': ({'id': 1575}, {'id': 1576})}}
- {'action': 'edit', 'id': 1574, 'index': 1, 'old': None, 'new': {'id': 1577}}
- {'action': 'create', 'value': {'id': 1578, 'type': 'IfcGridAxis', 'AxisTag': '01', 'AxisCurve': None, 'SameSense': True}}
- {'action': 'edit', 'id': 1560, 'index': 8, 'old': None, 'new': ({'id': 1578},)}
- {'action': 'create', 'value': {'id': 1579, 'type': 'IfcCartesianPoint', 'Coordinates': (0.0, -2.0)}}
- {'action': 'create', 'value': {'id': 1580, 'type': 'IfcCartesianPoint', 'Coordinates': (0.0, 22.0)}}
- {'action': 'create', 'value': {'id': 1581, 'type': 'IfcPolyline', 'Points': ({'id': 1579}, {'id': 1580})}}
- {'action': 'edit', 'id': 1578, 'index': 1, 'old': None, 'new': {'id': 1581}}
- {'action': 'create', 'value': {'id': 1582, 'type': 'IfcGridAxis', 'AxisTag': '02', 'AxisCurve': None, 'SameSense': True}}
- {'action': 'edit', 'id': 1560, 'index': 8, 'old': ({'id': 1578},), 'new': ({'id': 1578}, {'id': 1582})}
- {'action': 'create', 'value': {'id': 1583, 'type': 'IfcCartesianPoint', 'Coordinates': (10.0, -2.0)}}
- {'action': 'create', 'value': {'id': 1584, 'type': 'IfcCartesianPoint', 'Coordinates': (10.0, 22.0)}}
- {'action': 'create', 'value': {'id': 1585, 'type': 'IfcPolyline', 'Points': ({'id': 1583}, {'id': 1584})}}
- {'action': 'edit', 'id': 1582, 'index': 1, 'old': None, 'new': {'id': 1585}}
- {'action': 'create', 'value': {'id': 1586, 'type': 'IfcGridAxis', 'AxisTag': '03', 'AxisCurve': None, 'SameSense': True}}
- {'action': 'edit', 'id': 1560, 'index': 8, 'old': ({'id': 1578}, {'id': 1582}), 'new': ({'id': 1578}, {'id': 1582}, {'id': 1586})}
- {'action': 'create', 'value': {'id': 1587, 'type': 'IfcCartesianPoint', 'Coordinates': (20.0, -2.0)}}
- {'action': 'create', 'value': {'id': 1588, 'type': 'IfcCartesianPoint', 'Coordinates': (20.0, 22.0)}}
- {'action': 'create', 'value': {'id': 1589, 'type': 'IfcPolyline', 'Points': ({'id': 1587}, {'id': 1588})}}
- {'action': 'edit', 'id': 1586, 'index': 1, 'old': None, 'new': {'id': 1589}}
Traceback (most recent call last):
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\ifcopenshell\file.py", line 352, in undo
    transaction.rollback()
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\ifcopenshell\file.py", line 167, in rollback
    element = self.file.by_id(operation["id"])
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\ifcopenshell\file.py", line 510, in by_id
    return self[id]
           ~~~~^^^^
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\ifcopenshell\file.py", line 497, in __getitem__
    return entity_instance(self.wrapped_data.by_id(key), self)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\ifcopenshell\ifcopenshell_wrapper.py", line 8884, in by_id
    return _ifcopenshell_wrapper.file_by_id(self, id)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Instance #1565 not found

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\bonsai\bim\ifc.py", line 426, in callback
    tool.Ifc.get().undo()
  File "C:\Users\Andrej\AppData\Roaming\Blender Foundation\Blender\4.4\extensions\.local\lib\python3.11\site-packages\ifcopenshell\file.py", line 354, in undo
    raise UndoSystemError("Error during transaction undo.", transaction) from e
ifcopenshell.file.UndoSystemError: Error during transaction undo.
```

aothms committed:
python39 compatibility

Andrej730 committed:
ifcopenshell.file.Transaction - add static typing

Andrej730 committed:
Fix typo in 1619ae854

Andrej730 committed:
Fix resetting perspective camera on drawing activation (1619ae854)

Andrej730 committed:
Fix confused target_view callbacks (1619ae854)

Fix using the same function name twice leading to DocProperties using camera callback too.

Andrej730 committed:
typing

FrogsInSpace committed:
Update links and version info ifcmax.rst

Andrej730 committed:
Align new Model drawings to viewport by default

Previously it was creating not very useful top-down camera at `(0,0,0)`, now it should be more intuitive.

Example - https://imgur.com/a/GHrvCSJ

Andrej730 committed:
drawing.create_camera - remove setting location line that had no effect

Apparently it had no effect from the beginning in 074fe0d and lost any use in 6a8292d, when we started to base position on current cursor's location.

Andrej730 committed:
generate_drawing_matrix - simplify logic

theoryshaw committed:
fix #6683 - When placing `Level (Plan)` annotation, have it take on the 'z' dimension of the 3d cursor, verses the front plane of the drawing camera

brunoperdigao committed:
See #6332. Fix issue where imperial precision were being used by metric formatting.

brunoperdigao committed:
See #6536. Update cardinal point when editing `IfcMaterialProfileSetUsage`.

Moult committed:
Fix bug where appending assets could accidentally create duplicate contexts

Moult committed:
Fix bug in IFC4X3 when convering properties to quantities

Andrej730 committed:
Fix tests after specifying types more precisely in def9352ad

Andrej730 committed:
Fix drawing test after be657748f

Andrej730 committed:
Fix drawing styles update on underlay toggle after 9329390

Andrej730 committed:
drawing.prop - reuse generic pset update callback

Andrej730 committed:
typing

Andrej730 committed:
Remove 2 years old fallback for missing ShadingStyles prop in drawing's pset (38cf88d)

Andrej730 committed:
Fix IFC operators issues for drawing styles operations (38cf88d)

Andrej730 committed:
drawing.operator - reuse blender.get_view3d_space

Andrej730 committed:
bim.activate_drawing_style - show warnings if some property failed to be set

Otherwise errors were passing silently (just console errors) and remained unfixed. 'PYTEST_VERSION' is needed to be even more strict when we're running pytests, which allows us to catch outdated Blender properties in time.

Example - https://i.imgur.com/K7Trkxf.png

Andrej730 committed:
black .

falken10vdl committed:
system info copied when clicking in Bonsai Version (#6660)

Krande committed:
use rattler-build to speed up daily conda builds and fix deps (#6692)

* use rattler-build to speed up daily conda builds and add typing_extensions

* minor update

Krande committed:
Update ci-ifcopenshell-conda-daily.yml

Update conda ci

Krande committed:
Update ci-ifcopenshell-conda-daily.yml

strip daily conda build version date string

Krande committed:
separate osx-x86 and osx-arm64 for the conda daily builds (#6694)

* separate osx-x86 and osx-arm64 for the conda daily builds

* remove running this in PR. No need.

Krande committed:
Update ci-ifcopenshell-conda-daily.yml

fix missing comma

Krande committed:
Update ci-ifcopenshell-conda-daily.yml

try to get even more space on ubuntu ci runner

Krande committed:
Update ci-ifcopenshell-conda-daily.yml

remove dotnet to save space on ubuntu runner

Krande committed:
Update ci-ifcopenshell-conda-daily.yml

move rattler build output dir outside of build dir

Krande committed:
Update ci-ifcopenshell-conda-daily.yml

remove env var

Andrej730 committed:
Address PR review changes #6660

Andrej730 committed:
ci-black-formatting - print diff

falken10vdl committed:
Update type manager preview on updating paramteric door/window type (#6656)

Andrej730 committed:
typing

Andrej730 committed:
Mark thumbnails as outdated instead of updating them right away

As user may never open them in type manager

Andrej730 committed:
Drawing styles - remove deprecated properties / property values

Deprecated properties:
- scene.render.engine + BLENDER_EEVEE
- scene.render.simplify_shadows
- scene.render.simplify_shadows_render
- space.overlay.backwire_opacity
- space.overlay.show_edges
- space.overlay.show_occlude_wire
- scene.display.shading.wireframe_color_type + MATERIAL

sboddy committed:
Improve drawing activation as per #6653 (#6659)

Drawings are now activated in a bit different way - not by the camera button located above the drawings list, but by using the camera indicator next to each drawing. Which allows do change active drawing very quickly. All activation options (shift+, alt+) are also available from this indicator.

Now - https://imgur.com/a/DmnWkul

Before - https://imgur.com/a/zjPDkPa

RickBrice committed:
Adds object placement to stationing referent.

brunoperdigao committed:
See #6537. Fix object display type when exiting aggregate mode.

Moult committed:
Accommodate invalid missing mapped representations and items coming from Tekla (2023)

Andrej730 committed:
Prevent class assignment in non-object mode #6697

Example - https://i.imgur.com/KQYQLQZ.png

Andrej730 committed:
black .

Andrej730 committed:
Geometry changes will be lost - show warning only for IFC objects

Previously it would also occur for vanilla Blender objects not connected to IFC.

Andrej730 committed:
typing

Andrej730 committed:
Add notes on Tekla accommodations (7c3870595)

hannesw committed:
Fixed typo in __init__.py

See https://github.com/IfcOpenShell/IfcOpenShell/issues/6625

Andrej730 committed:
bim.show_openings - expand description

Andrej730 committed:
bim.show_opening - simplify code

Andrej730 committed:
IfcImporter.create_generic_elements - prevent argument mutation

as it may have unexpected consequences

Andrej730 committed:
Skip IfcOpeningElements arrays during base import, setup constraint on opening load #6688

Andrej730 committed:
util.get_aggregate - make it more readable (5597095)

Andrej730 committed:
Use common method for getting active drawing entity

Andrej730 committed:
drawing.operator not to use DrawingsData

Andrej730 committed:
typing

Andrej730 committed:
drawing styles - add descriptions to render types, remove NONE type

NONE type removed as it was unused and basically meant the same as 'VIEWPORT'.

Example - https://i.imgur.com/7r6CJdg.png

Andrej730 committed:
common method for getting active drawing style

Andrej730 committed:
Remove drawing styles attributes

It's been awhile since they were unused (2c301d7) and now they're replaced with EPset_Drawing.Metadata

Andrej730 committed:
draw_custom_context_menu - small refactor to improve readibility

aothms committed:
Align preprocessor flags #6706

Andrej730 committed:
util.doc - statically type json structures

brunoperdigao committed:
Fix aggregate decorator after #6537.

brunoperdigao committed:
See #6536. Recalculate profile only for selected objects.

Andrej730 committed:
Reapply 06f1477 to fix not appearing IFC Description button

Not that skipping non-string properties was a bad idea, but inserting it right at the beginning was - non-string IFC properties were missing IFC descriptions because of it. Moved it past `Attribute` check to avoid this issue.

For anyone hearing about this feature for the first time - https://imgur.com/a/dNRao3R

I assume this was also the issue that was encountered in #5892

Andrej730 committed:
EPset_Drawing.Metadata - add description

See https://i.imgur.com/0fR9fRj.png

Andrej730 committed:
Remove drawing styles attributes prop (a50c633cf)

Andrej730 committed:
typing

Andrej730 committed:
Fix tests missing assert statement - add assert and fix them

Andrej730 committed:
Annotator.add_text - remove ancient dead code

It was deprecated a while ago in 43869c3, when we stopped using curves for text annotations. bim.resize_text was also using curve annotations and now also was unused

Andrej730 committed:
Allow adding arrays to openings and spatial elements 0b8fb41 0e37528

Andrej730 committed:
Remove unnecessary IFC operators

Andrej730 committed:
Add a test for be65774

Andrej730 committed:
bim.add_drawing - document camera placement behaviour in description

See https://i.imgur.com/VfOjGHk.png

brunoperdigao committed:
Fix polyline tool calculation when dealing with 180 degrees angles.

brunoperdigao committed:
See #6208. Change wall merge tolerance for more consistent behavior.

brunoperdigao committed:
Fix polyline plane method decorator to show properly.
The `plane_method` where not being cleared after canceling or confirming
the polyline operator.

theoryshaw committed:
A preference to list those classes that are not affected by the 'cut decorator'.

Andrej730 committed:
pyproject.toml - enableTypeIgnoreComments=false

Andrej730 committed:
SvgWriter - make required arguments more explicit

Andrej730 committed:
black .

Andrej730 committed:
bim.split_along_edge - improve descriptions, add error handling

Andrej730 committed:
typing

Andrej730 committed:
operator.drawing - remove unnecessary ifc operators

Andrej730 committed:
Fix error using bim.updating_representation for object with material set

Possibly occurred after 4c255cc, but haven't investigated deeper.

@BIMvoice
The error you were reporting about Split Along Edge operator was caused by this issue, should be fixed now too.
See https://jmp.sh/hgJEHsgH

Example error traceback:
```
Error: Python: Traceback (most recent call last):
  File "C\bonsai\bim\ifc.py", line 487, in execute_ifc_operator
    result = getattr(operator, "_execute")(context)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C\bonsai\bim\module\geometry\operator.py", line 541, in _execute
    self.update_obj_mesh_representation(context, obj)
  File "C\bonsai\bim\module\geometry\operator.py", line 666, in update_obj_mesh_representation
    tool.Ifc.link(new_representation, data)
  File "C\bonsai\tool\ifc.py", line 230, in link
    IfcStore.link_element(element, obj)
  File "C\bonsai\bim\ifc.py", line 258, in link_element
    tool.Geometry.get_mesh_props(obj).ifc_definition_id = element.id()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C\bonsai\tool\geometry.py", line 92, in get_mesh_props
    return mesh.BIMMeshProperties
           ^^^^^^^^^^^^^^^^^^^^^^
ReferenceError: StructRNA of type Mesh has been removed
```

steverugi commented:
Hi @falken10vdl 
please any udpate on your PR? it's something I'd really love to see implemented
thanks for your contribution, much appreciated

falken10vdl commented:
Hi @steverugi . I guess the core developers are taking a look to it. I will be glad to update whatever they think is needed.
In the meantime if you want you can use it by installing the dev environment and pulling it. Find attached the documentation to get that installed in you windows pc (step 15 is exactly what you are after). Maybe you can try it and provide feedback if something is not working for you ( https://github.com/IfcOpenShell/IfcOpenShell/pull/6494  ). 
Cheers!
[Windows - Bonsai 0.8.2 IDE QS.zip](https://github.com/user-attachments/files/20263540/Windows.-.Bonsai.0.8.2.IDE.QS.zip)
[windows_files.zip](https://github.com/user-attachments/files/20263551/windows_files.zip)

Moult committed:
Fix #6669. bSDD now supports test dictionaries, refactor domain to dictionary, support zero length keyword searches, multi-dict support, and basic tests.

Feature tests can now have more sophisticated per UIList item spies.

aothms committed:
ifdefs around geometry handling for 4.3 rc schemas #6710

Moult committed:
More bSDD tests to cover all current functionality

Moult committed:
See #4699. Continue refactoring logic from drawing decorator into DecoratorData.

Moult committed:
Update test cases to reflect new drawing activation UI

Moult committed:
Update colour by property test to reflect new behaviour of only showing geometry not empties

ALL (feature) TESTS PASS YESS

aothms committed:
OcctNoCleanTriangulation CacheShapes PermissiveShapeReuse setting; Cgal conditional copy during triangulation; occt cheaper check for triangulation existance #6712

falken10vdl commented:
Sorry. I just tried to rebase with vscode to the current base branch and it looks it has pushed everything in between instead of applying the changes to the fresh base. Clean in https://github.com/IfcOpenShell/IfcOpenShell/pull/6717
